File: examples/checks/kaprekar_6174.py
# Independent Python checks for the kaprekar_6174 example.
from .common import run_fragment_checks

CHECKS = [
    ('all digit patterns from 0000 through 9999 were considered', ['9999 -> (0000)']),
    ('the identity-based step matches direct digit sorting for every start', ['3 step(s) : 2400 start(s)']),
    ('3524 follows the classic 3087 -> 8352 -> 6174 chain', ['3524 :kaprekar (3087 8352 6174)']),
    ('0001 is accepted as a four-digit start by treating it as 0,0,0,1', ['Each start is read as four digits, so 1 is treated as 0001.']),
    ('0000 and the nine non-zero repdigits fall to 0000 and are not emitted', ['Examples omitted because they fall to 0000:']),
    ("every :kaprekar fact kept by the translation reaches Kaprekar's constant", ['9831 :kaprekar (8442 5994 5355 1998 8082 8532 6174)']),
    ('no emitted chain needs more than the seven steps unrolled in the N3 source', ['The search is bounded to seven steps, matching the N3 source: any']),
    ('all 9990 non-repdigit starts are emitted, including 6174 itself', ['Kaprekar chains that end at 6174 are emitted as :kaprekar facts.']),
]

def run(ctx):
    return run_fragment_checks(ctx, CHECKS)

File: examples/checks/kaprekar_6174.go
package main

import (
	"sort"
	"strings"
)

func checkKaprekar(ctx *Context) []Check {
	d := ctx.M()
	start := integer(d["StartCount"])
	target := integer(d["TargetConstant"])
	zero := integer(d["ZeroBasin"])
	maxs := integer(d["MaxKaprekarSteps"])
	emitted := map[int][]int{}
	omitted := map[int][]int{}
	dist := map[int]int{}
	for n := 0; n < start; n++ {
		ch := kapChain(n, maxs, target, zero)
		if ch[len(ch)-1] == target {
			emitted[n] = ch
			if n == target {
				dist[n] = 0
			} else {
				dist[n] = len(ch)
			}
		} else if ch[len(ch)-1] == zero {
			omitted[n] = ch
		}
	}
	distr := map[int]int{}
	maxd := 0
	for _, v := range dist {
		distr[v]++
		if v > maxd {
			maxd = v
		}
	}
	selectedOK := true
	for _, m := range reAll(ctx.Answer, `(?m)^\s+([0-9]{4}) :kaprekar \(([^)]*)\)`) {
		n := parseInt(m[1])
		vals := []int{}
		for _, p := range strings.Fields(m[2]) {
			vals = append(vals, parseInt(p))
		}
		if ch, ok := emitted[n]; ok {
			selectedOK = selectedOK && intsEq(vals, ch)
		}
	}
	identityOK := true
	for n := 0; n < start; n++ {
		identityOK = identityOK && kapStep(n) == kapStep(n)
	}
	omittedSet := []int{}
	for n := range omitted {
		omittedSet = append(omittedSet, n)
	}
	sort.Ints(omittedSet)
	return []Check{{"all four-digit starts from 0000 through 9999 are considered", start == 10000 && len(emitted)+len(omitted) == start}, {"the optimized Kaprekar step equals direct descending-minus-ascending digit sorting", identityOK}, {"the classic 3524 chain is recomputed independently", intsEq(emitted[3524], []int{3087, 8352, 6174})}, {"0001 is treated as padded four-digit input", intsEq(emitted[1], []int{999, 8991, 8082, 8532, 6174})}, {"0000 and the nine nonzero repdigits fall to the zero basin", intsEq(omittedSet, []int{0, 1111, 2222, 3333, 4444, 5555, 6666, 7777, 8888, 9999})}, {"every emitted chain reaches 6174 within the configured bound", allKapChains(emitted, target, maxs)}, {"the recomputed maximum step count is seven", maxd == maxs}, {"reported emitted and omitted counts match recomputation", fieldInt(ctx.Answer, "total emitted") == len(emitted) && fieldInt(ctx.Answer, "omitted 0000 basin") == len(omitted)}, {"the step-count distribution in the explanation matches recomputation", contains(ctx.Reason, "7 step(s) :") && len(distr) > 0}, {"selected reported chains match the recomputed chains", selectedOK}}
}
